# 01. Wczytanie danych i wstępna analiza


In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("../data/hotel_bookings_updated_2024.csv")

In [3]:
df.head()

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,city
0,Resort Hotel - Chandigarh,0,342,2024,July,30,27,0,0,2,...,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2024-07-27 22:16:40.916332324,Chandigarh
1,Resort Hotel - Mumbai,0,737,2024,April,17,28,0,0,2,...,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2024-04-28 21:56:21.507509066,Mumbai
2,Resort Hotel - Delhi,0,7,2024,September,37,10,0,1,1,...,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2024-09-10 03:46:25.734029096,Delhi
3,Resort Hotel - Kolkata,0,13,2024,August,33,14,0,1,1,...,304.0,NaN,0,Transient,75.0,0,0,Check-Out,2024-08-14 18:07:10.049669568,Kolkata
4,Resort Hotel - Lucknow,0,14,2024,September,37,14,0,2,2,...,240.0,NaN,0,Transient,98.0,0,1,Check-Out,2024-09-14 14:27:32.473846000,Lucknow


In [4]:
df.shape

(119390, 33)

In [5]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 33 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  str    
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  str    
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal                       

In [6]:
df.describe()

,is_canceled,lead_time,arrival_date_year,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,children,babies,is_repeated_guest,previous_cancellations,previous_bookings_not_canceled,booking_changes,agent,company,days_in_waiting_list,adr,required_car_parking_spaces,total_of_special_requests
count,119390.000000,119390.000000,119390.0,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,119386.000000,119390.000000,119390.000000,119390.000000,119390.000000,119390.000000,103050.000000,6797.000000,119390.000000,119390.000000,119390.000000,119390.000000
mean,0.370416,104.011416,2024.0,26.375835,15.723394,0.927599,2.500302,1.856403,0.103890,0.007949,0.031912,0.087118,0.137097,0.221124,86.693382,189.266735,2.321149,101.831122,0.062518,0.571363
std,0.482918,106.863097,0.0,15.021596,8.805079,0.998613,1.908286,0.579261,0.398561,0.097436,0.175767,0.844336,1.497437,0.652306,110.774548,131.655015,17.594721,50.535790,0.245291,0.792798
min,0.000000,0.000000,2024.0,1.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,6.000000,0.000000,-6.380000,0.000000,0.000000
25%,0.000000,18.000000,2024.0,13.000000,8.000000,0.000000,1.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,9.000000,62.000000,0.000000,69.290000,0.000000,0.000000
50%,0.000000,69.000000,2024.0,26.000000,16.000000,1.000000,2.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,14.000000,179.000000,0.000000,94.575000,0.000000,0.000000
75%,1.000000,160.000000,2024.0,39.000000,23.000000,2.000000,3.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,229.000000,270.000000,0.000000,126.000000,0.000000,1.000000
max,1.000000,737.000000,2024.0,52.000000,31.000000,19.000000,50.000000,55.000000,10.000000,10.000000,1.000000,26.000000,72.000000,21.000000,535.000000,543.000000,391.000000,5400.000000,8.000000,5.000000


In [7]:
df.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

## Wnioski z analizy ilości braków w zbiorze danych

W zbiorze danych występują brakujące wartości w czterech kolumnach:
- `children` – 4 brakujące wartości,
- `country` – 488 brakujących wartości,
- `agent` – 16 340 brakujących wartości,
- `company` – 112 593 brakujące wartości.



In [8]:
df.columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'adults', 'children', 'babies', 'meal',
       'country', 'market_segment', 'distribution_channel',
       'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type', 'agent',
       'company', 'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests',
       'reservation_status', 'reservation_status_date', 'city'],
      dtype='str')

In [9]:
df.dtypes

hotel                                 str
is_canceled                         int64
lead_time                           int64
arrival_date_year                   int64
arrival_date_month                    str
arrival_date_week_number            int64
arrival_date_day_of_month           int64
stays_in_weekend_nights             int64
stays_in_week_nights                int64
adults                              int64
children                          float64
babies                              int64
meal                                  str
country                               str
market_segment                        str
distribution_channel                  str
is_repeated_guest                   int64
previous_cancellations              int64
previous_bookings_not_canceled      int64
reserved_room_type                    str
assigned_room_type                    str
booking_changes                     int64
deposit_type                          str
agent                             

In [10]:
df["company"].nunique()

352

In [11]:
df["agent"].nunique()

333

In [12]:
df.duplicated().sum()

np.int64(0)


W zbiorze danych nie wykryto żadnych duplikatów, dlatego nie ma konieczności ich usuwania.

In [13]:
cancel_percent = (df["is_canceled"].value_counts(normalize=True) * 100).round(2)

print(f"Nieanulowane rezerwacje: {cancel_percent[0]}%")
print(f"Anulowane rezerwacje: {cancel_percent[1]}%")

Nieanulowane rezerwacje: 62.96%
Anulowane rezerwacje: 37.04%


In [14]:
missing_percent = (df.isnull().sum() / len(df) * 100).round(2)

print("Kolumny zawierające brakujące wartości:\n")

for column, percent in missing_percent.items():
    if percent > 0:
        print(f"{column}: {percent}%")

Kolumny zawierające brakujące wartości:

country: 0.41%
agent: 13.69%
company: 94.31%


In [15]:
df["company"].value_counts(dropna=False)

company
NaN      112593
40.0        927
223.0       784
67.0        267
45.0        250
          ...  
479.0         1
489.0         1
229.0         1
481.0         1
497.0         1
Name: count, Length: 353, dtype: int64

In [16]:
df.groupby(df["company"].notnull())["is_canceled"].mean()

company
False    0.382200
True     0.175224
Name: is_canceled, dtype: float64

Kolumna "company" zawiera 94,31% brakujących wartości. Dodatkowo przechowuje jedynie identyfikatory firm, bez informacji opisowych. Pomimo tego, że rezerwacje firmowe były rzadziej anulowane, bardzo wysoki odsetek braków oraz brak dodatkowych informacji o firmach sprawiają, że kolumna nie będzie wykorzystywana w dalszej analizie

In [17]:
df.drop(columns=["company"], inplace=True)

In [18]:
df.columns


Index(['hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'adults', 'children', 'babies', 'meal',
       'country', 'market_segment', 'distribution_channel',
       'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type', 'agent',
       'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests',
       'reservation_status', 'reservation_status_date', 'city'],
      dtype='str')

In [19]:
missing = df.isnull().sum()
missing[missing > 0]

children        4
country       488
agent       16340
dtype: int64

In [20]:
df[df["children"].isnull()]

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date,city
40600,City Hotel - Bangalore,1,2,2024,June,24,16,1,0,2,...,No Deposit,NaN,0,Transient-Party,12.0,0,1,Canceled,2024-06-16 15:52:52.543534160,Bangalore
40667,City Hotel - Mumbai,1,1,2024,July,28,14,0,2,2,...,No Deposit,14.0,0,Transient-Party,12.0,0,1,Canceled,2024-07-14 03:17:23.101123218,Mumbai
40679,City Hotel - Hyderabad,1,1,2024,September,39,26,0,2,3,...,No Deposit,NaN,0,Transient-Party,18.0,0,2,Canceled,2024-09-26 12:54:34.311703756,Hyderabad
41160,City Hotel - Pune,1,8,2024,September,38,21,2,5,2,...,No Deposit,9.0,0,Transient-Party,76.5,0,1,Canceled,2024-09-21 03:33:06.786052316,Pune


In [21]:
df[df["children"].isnull()][["children", "adults", "babies", "country", "agent"]]

,children,adults,babies,country,agent
40600,NaN,2,0,PRT,NaN
40667,NaN,2,0,PRT,14.0
40679,NaN,3,0,PRT,NaN
41160,NaN,2,0,PRT,9.0


In [22]:
df.dropna(subset=["children"], inplace=True)

In [23]:
missing = df.isnull().sum()
missing[missing > 0]

country      488
agent      16338
dtype: int64

W kolumnie children znaleziono jedynie 4 brakujące wartości (0,003% wszystkich rekordów). Z uwagi na ich bardzo małą liczbę oraz brak możliwości wiarygodnego uzupełnienia, rekordy te zostały usunięte z dalszej analizy.

In [24]:
df[df["country"].isnull()][
    ["hotel", "country", "market_segment", "distribution_channel", "agent"]
]

,hotel,country,market_segment,distribution_channel,agent
30,Resort Hotel - Kochi,NaN,Direct,Direct,NaN
4127,Resort Hotel - Chandigarh,NaN,Offline TA/TO,TA/TO,NaN
7092,Resort Hotel - Pune,NaN,Corporate,Corporate,NaN
7860,Resort Hotel - Delhi,NaN,Direct,Direct,NaN
8779,Resort Hotel - Chandigarh,NaN,Corporate,Corporate,NaN
...,...,...,...,...,...
65908,City Hotel - Kolkata,NaN,Complementary,Corporate,NaN
65909,City Hotel - Lucknow,NaN,Complementary,Corporate,NaN
65910,City Hotel - Kolkata,NaN,Complementary,Corporate,NaN
80830,City Hotel - Indore,NaN,Groups,TA/TO,37.0


In [25]:
df[df["country"].isnull()]["hotel"].value_counts()

hotel
Resort Hotel - Pune          41
Resort Hotel - Hyderabad     38
Resort Hotel - Kolkata       37
Resort Hotel - Kochi         34
Resort Hotel - Bhopal        34
Resort Hotel - Delhi         33
Resort Hotel - Ahmedabad     32
Resort Hotel - Indore        30
Resort Hotel - Lucknow       29
Resort Hotel - Chandigarh    28
Resort Hotel - Goa           28
Resort Hotel - Chennai       28
Resort Hotel - Jaipur        27
Resort Hotel - Bangalore     25
Resort Hotel - Mumbai        20
City Hotel - Mumbai           4
City Hotel - Delhi            3
City Hotel - Pune             3
City Hotel - Ahmedabad        2
City Hotel - Chandigarh       2
City Hotel - Hyderabad        2
City Hotel - Kolkata          2
City Hotel - Jaipur           1
City Hotel - Goa              1
City Hotel - Kochi            1
City Hotel - Lucknow          1
City Hotel - Indore           1
City Hotel - Bangalore        1
Name: count, dtype: int64

In [26]:
df.groupby(df["country"].isnull())["is_canceled"].mean()

country
False    0.371352
True     0.137295
Name: is_canceled, dtype: float64

In [27]:
df.dropna(subset=["country"], inplace=True)

In [28]:
missing = df.isnull().sum()
missing[missing > 0]

agent    16004
dtype: int64

In [29]:
df["agent"].value_counts()

agent
9.0      31959
240.0    13871
1.0       7191
14.0      3638
7.0       3539
         ...  
444.0        1
408.0        1
388.0        1
453.0        1
480.0        1
Name: count, Length: 332, dtype: int64

In [30]:
df.groupby(df["agent"].isnull())["is_canceled"].mean()

agent
False    0.390314
True     0.249438
Name: is_canceled, dtype: float64

In [31]:
df[df["agent"].isnull()][["market_segment", "distribution_channel"]].value_counts()

market_segment  distribution_channel
Direct          Direct                  5631
Corporate       Corporate               3941
Groups          TA/TO                   2024
                Direct                  1076
                Corporate               1045
Complementary   Direct                   540
Offline TA/TO   TA/TO                    381
Corporate       TA/TO                    319
Online TA       TA/TO                    243
Aviation        Corporate                204
Direct          TA/TO                    162
Corporate       Direct                   159
Complementary   Corporate                 73
Direct          Corporate                 71
Online TA       Direct                    70
                Corporate                 31
Complementary   TA/TO                     18
Aviation        TA/TO                      8
Offline TA/TO   Direct                     4
Direct          Undefined                  1
Offline TA/TO   Corporate                  1
Direct          GD

In [32]:
df["agent"] = df["agent"].fillna(0)

In [33]:
missing = df.isnull().sum()
missing[missing > 0]


Series([], dtype: int64)

Podsumowanie czyszczenia danych
Usunięto kolumnę company, ponieważ zawierała ponad 94% brakujących wartości.
Usunięto 4 rekordy z brakującą wartością w kolumnie children (0,003% danych).
Usunięto 488 rekordów z brakującą wartością w kolumnie country (0,41% danych), ponieważ nie było możliwości ich wiarygodnego uzupełnienia.
W kolumnie agent brakujące wartości zastąpiono wartością 0, ponieważ analiza wykazała, że NaN najprawdopodobniej oznacza brak pośrednika, a nie brak danych.

In [34]:
df.shape


(118898, 32)

In [35]:
df.info()

<class 'pandas.DataFrame'>
Index: 118898 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           118898 non-null  str    
 1   is_canceled                     118898 non-null  int64  
 2   lead_time                       118898 non-null  int64  
 3   arrival_date_year               118898 non-null  int64  
 4   arrival_date_month              118898 non-null  str    
 5   arrival_date_week_number        118898 non-null  int64  
 6   arrival_date_day_of_month       118898 non-null  int64  
 7   stays_in_weekend_nights         118898 non-null  int64  
 8   stays_in_week_nights            118898 non-null  int64  
 9   adults                          118898 non-null  int64  
 10  children                        118898 non-null  float64
 11  babies                          118898 non-null  int64  
 12  meal                            

In [36]:
df.to_csv("../data/hotel_bookings_clean.csv", index=False)

In [37]:
import os

print(os.path.exists("../data/hotel_bookings_clean.csv"))

True
